src/v/cloud_storage/tx_range_manifest.cc

```diff
@@ -12,20 +12,215 @@
+    bool Bool(bool) { return false; }
+    bool Int(int i) { return Int64(i); }
+    bool Uint(unsigned u) { return Int64(u); }
+    bool Int64(int64_t i64) {
+        switch (_state) {
+        case state::expect_version_value:
+            version = static_cast<int>(i64);
+            _state = state::expect_compat_version_key;
+            return true;
+        case state::expect_compat_version_value:
+            compat_version = static_cast<int>(i64);
+            if (
+              compat_version
+              > static_cast<int>(tx_range_manifest_version::current_version)) {
+                throw std::runtime_error(fmt::sprintf(
+                  "Can't deserialize tx manifest, supported version {}, "
+                  "manifest "
+                  "version {}, compatible version {}",
+                  static_cast<int32_t>(
+                    tx_range_manifest_version::current_version),
+                  version,
+                  compat_version));
+            }
+            _state = state::expect_ranges_key;
+            return true;
+        case state::expect_range_pid_id_value:
+            ranges.back().pid.id = model::producer_id(i64);
+            _state = state::expect_range_pid_epoch_key;
+            return true;
+        case state::expect_range_pid_epoch_value:
+            if (
+              i64 < std::numeric_limits<int16_t>::min()
+              || i64 > std::numeric_limits<int16_t>::max()) {
+                return false;
+            }
+            ranges.back().pid.epoch = model::producer_epoch(
+              static_cast<int16_t>(i64));
+            _state = state::expect_range_first_key;
+            return true;
+        case state::expect_range_first_value:
+            ranges.back().first = model::offset(i64);
+            _state = state::expect_range_last_key;
+            return true;
+        case state::expect_range_last_value:
+            ranges.back().last = model::offset(i64);
+            _state = state::expect_range_end_object;
+            return true;
+        default:
+            return false;
+        }
+    }
+    bool Uint64(uint64_t u64) {
+        if (u64 > std::numeric_limits<int64_t>::max()) {
+            return false;
+        } else {
+            return Int64(static_cast<int64_t>(u64));
+        }
+    }
+    bool Double(double) { return false; }
+    /// enabled via kParseNumbersAsStringsFlag, string is not null-terminated
+    /// (use length)
+    bool
+    RawNumber(const Ch* /*str*/, rapidjson::SizeType /*len*/, bool /*copy*/) {
+        return false;
+    }
+    bool String(const Ch*, rapidjson::SizeType, bool) { return false; }
+    bool StartObject() {
+        switch (_state) {
+        case state::expect_manifest_object:
+            _state = state::expect_version_key;
+            return true;
+        case state::expect_ranges_entry_or_end_array:
+            ranges.push_back(model::tx_range());
+            _state = state::expect_range_pid_id_key;
```

Comment:
Looks like the implementation expects certain order of keys. This is not necessary. You can expect some key and then based on the value of the key you can make the decision to expect pid or range or whatever.